### PR TITLE
debug: add yappi, callgrind output support

### PR DESCRIPTION
**Pull request by skshetry**
#### How to use:
0. `pip install yappi`
1. add `--yappi` at the end of any command.
2. Install `kcachegrind`/`qcachegrind`.
3. Open the file, which is in the form of `callgrind.dvc-{id}.out`format with `{k,q}cachegrind`.

I haven't really used this that much, so you'd have to see how this works. :)

Related: https://github.com/iterative/dvc/issues/6627 (not closing it yet, let's see how this works).

Screen Shot 2021-09-21 at 22 20 34

### CODE REVIEW

**Comment by efiop** [comment]
Side note: would be amazing to support these flags in `dvc-bench`. https://github.com/iterative/dvc-bench/issues/301

**Review by efiop — APPROVED** [review]
Great stuff! 🔥

File: dvc/_debug.py [hunk 1]
```diff
@@ -5,6 +5,34 @@
...
+    yappi.set_clock_type("wall" if wall_clock else "cpu")
+
+    yappi.start()
+    yield
+    yappi.stop()
+
+    if save_output:
+        stats = yappi.get_func_stats()
+        uid = time.strftime("%Y%m%d_%H%M%S")
+        stats.save(f"callgrind.dvc-{uid}.out", "callgrind")
```
Comment (skshetry) [inline dvc/_debug.py:28]: Went autogenerated way with this yappi, let me know if you want to use `--cprofile-dump dump.prof` like flags 😄.

**Comment by skshetry** [comment]
> Let's add to https://github.com/iterative/dvc/wiki/Debugging-DVC ?

Working on it. :)

File: dvc/_debug.py (hunk 1 above)
Comment (efiop) [reply dvc/_debug.py:28]: Might need that option for https://github.com/iterative/dvc-bench/issues/301 , so as long as we could introduce it later - it is fine.

**Comment by skshetry** [comment]
Btw, we should probably have debug builds for binaries.

File: dvc/_debug.py (hunk 1 above)
Comment (skshetry) [reply dvc/_debug.py:28]: Let me work on that then. :)

**Comment by dberenbaum** [comment]
Any example output you can show? Or is that part of adding it to the Wiki?

**Comment by skshetry** [comment]
> Any example output you can show? Or is that part of adding it to the Wiki?

Added example output (it is part of kcachegrind though). There's not much to this than adding `--yappi` to the command, and then later opening the `callgrind.*` file through kcachegrind.

File: dvc/_debug.py [hunk 2]
```diff
@@ -70,6 +99,13 @@ def debugtools(args: "Namespace" = None, **kwargs):
             stack.enter_context(profile(kw.get("cprofile_dump")))
         if kw.get("instrument") or kw.get("instrument_open"):
             stack.enter_context(instrument(kw.get("instrument_open", False)))
+        if kw.get("yappi"):
+            from datetime import datetime
+
+            output = "callgrind.dvc-{0:%Y%m%d}_{0:%H%M%S}.out"
+            stack.enter_context(
+                yappi_profile(path=lambda: output.format(datetime.now()))
```
Comment (efiop) [inline dvc/_debug.py:107]:
```suggestion
                yappi_profile(path_factory=lambda: output.format(datetime.now()))
```
?

**Review by efiop — APPROVED** [review]
🔥 

These options are hidden anyway, so we can tweak them as we go.

File: dvc/_debug.py (hunk 2 above)
Comment (skshetry) [reply dvc/_debug.py:107]: yeah, we could do that too, but I am optimizing the API here. Note that this `yappi_profile` (along with all the others) can be used as a decorator and a context manager on any functions that you want to profile:

```python
c = count(start=1)


@yappi_profile(path=lambda: next(c))
def _diff(*args, **kwargs):
    pass
```

**Pull request merged**